vm/vm.go

```diff
@@ -0,0 +1,141 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vm
+
+import (
+	"fmt"
+	"math/big"
+
+	"C"
+
+	"github.com/bytecodealliance/wasmtime-go"
+
+	"github.com/onflow/cadence/runtime/interpreter"
+)
+
+type VM interface {
+	Invoke(name string, arguments ...interpreter.Value) (interpreter.Value, error)
+}
+
+type vm struct {
+	instance *wasmtime.Instance
+}
+
+func (m *vm) Invoke(name string, arguments ...interpreter.Value) (interpreter.Value, error) {
+	f := m.instance.GetExport(name).Func()
+
+	rawArguments := make([]interface{}, len(arguments))
+	for i, argument := range arguments {
+		rawArguments[i] = argument
+	}
+
+	res, err := f.Call(rawArguments...)
+	if err != nil {
+		return nil, err
+	}
+
+	if res == nil {
+		return nil, nil
+	}
+
+	return res.(interpreter.Value), nil
+}
+
+func NewVM(wasm []byte) (VM, error) {
+
+	config := wasmtime.NewConfig()
+	config.SetWasmReferenceTypes(true)
+
+	engine := wasmtime.NewEngineWithConfig(config)
+
+	store := wasmtime.NewStore(engine)
+
+	module, err := wasmtime.NewModule(store.Engine, wasm)
+	if err != nil {
+		return nil, err
+	}
+
+	var mem *wasmtime.Memory
+
+	intFunc := wasmtime.WrapFunc(store, func(offset int32, length int32) (interface{}, *wasmtime.Trap) {
+		if offset < 0 {
+			return nil, wasmtime.NewTrap(store, fmt.Sprintf("Int: invalid offset: %d", offset))
+		}
+
+		if length < 2 {
+			return nil, wasmtime.NewTrap(store, fmt.Sprintf("Int: invalid length: %d", length))
+		}
+
+		bytes := C.GoBytes(mem.Data(), C.int(length))
+
+		value := new(big.Int).SetBytes(bytes[1:])
+		if bytes[0] == 0 {
+			value = value.Neg(value)
+		}
+
+		return interpreter.NewIntValueFromBigInt(value), nil
+	})
+
+	stringFunc := wasmtime.WrapFunc(store, func(offset int32, length int32) (interface{}, *wasmtime.Trap) {
+		if offset < 0 {
+			return nil, wasmtime.NewTrap(store, fmt.Sprintf("String: invalid offset: %d", offset))
+		}
+
+		if length < 0 {
+			return nil, wasmtime.NewTrap(store, fmt.Sprintf("String: invalid length: %d", length))
+		}
+
+		bytes := C.GoBytes(mem.Data(), C.int(length))
+
+		return interpreter.NewStringValue(string(bytes)), nil
+	})
+
+	addFunc := wasmtime.WrapFunc(store, func(left, right interface{}) (interface{}, *wasmtime.Trap) {
+		leftNumber, ok := left.(interpreter.NumberValue)
+		if !ok {
+			return nil, wasmtime.NewTrap(store, fmt.Sprintf("add: invalid left: %#+v", left))
+		}
+
+		rightNumber, ok := right.(interpreter.NumberValue)
+		if !ok {
+			return nil, wasmtime.NewTrap(store, fmt.Sprintf("add: invalid right: %#+v", right))
+		}
+
+		return leftNumber.Plus(rightNumber), nil
+	})
+
+	instance, err := wasmtime.NewInstance(
+		store,
+		module,
+		[]*wasmtime.Extern{
+			intFunc.AsExtern(),
+			stringFunc.AsExtern(),
+			addFunc.AsExtern(),
```

Comment:
out of curiosity: when the imports are added to these extern functions (referring to [addRuntimeImports](https://github.com/onflow/cadence/pull/448/files#diff-1f6a4a5ec36bcf2c32f34ef4341b81af60a52fe5ebb2eb06981d5ed3bf2fbd4eR209-R213)), how does it map that import to the relevant function definition? I guess its by the order defined here and  [there](https://github.com/onflow/cadence/pull/448/files#diff-1f6a4a5ec36bcf2c32f34ef4341b81af60a52fe5ebb2eb06981d5ed3bf2fbd4eR209-R213)? (and both places needs to ensure the same order?)

Reply:
The mapping is by index, which is not great. You're right, the code generator declares imports, and wasmtime expects them to be passed in in the same order. 

I found this odd, because inside the WebAssembly module the imports are also named, and other WebAssembly APIs, [like the JavaScript one available in browsers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiate), provide means to pass imports by name.

Maybe wasmtime will also provide support for passing imports by name and matching them with the module's imports at some point